packages/framework/presence/src/presenceDatastoreManager.ts

```diff
@@ -200,14 +233,95 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+		// It is OK to queue all incoming messages as long as when we send, we send the queued data.
+		this.queuedData = mergeGeneralDatastoreMessageContent(this.queuedData, data);
+
+		const { allowableUpdateLatencyMs } = options;
+		const now = Date.now();
+		const thisMessageDeadline = now + allowableUpdateLatencyMs;
+
+		if (
+			// If the deadline for this message is later than the overall send deadline, then
+			// we may be able to exit early if a timer will take care of sending it.
+			thisMessageDeadline >= this.timer.expireTime &&
+			// If the timer has not expired, we can short-circuit because the timer will fire
+			// and cover this update. In other words, queuing this will be fast enough to
+			// meet its deadline, because a timer is already scheduled to fire before its deadline.
+			!this.timer.hasExpired()
+		) {
+			return;
+		}
+
+		// Either we need to send this message immediately, or we need to schedule a timer
+		// to fire at the send deadline that will take care of it.
+
+		// Note that timeoutInMs === allowableUpdateLatency, but the calculation is done this way for clarity.
+		const timeoutInMs = thisMessageDeadline - now;
+		const scheduleForLater = timeoutInMs > 0;
+
+		if (scheduleForLater) {
+			// Schedule the queued messages to be sent at the updateDeadline
+			this.timer.setTimeout(this.sendQueuedMessage.bind(this), timeoutInMs);
+		} else {
+			this.sendQueuedMessage();
+		}
+	}
+
+	/**
+	 * Send any queued signal immediately. Does nothing if no message is queued.
+	 */
+	private sendQueuedMessage(): void {
+		if (this.queuedData === undefined) {
+			return;
+		}
+
+		// Check for connectivity before sending updates.
+		if (!this.runtime.connected) {
+			// Clear the queued data since we're disconnected. We don't want messages
+			// to queue infinitely while disconnected.
+			this.queuedData = undefined;
+			return;
+		}
+
+		const clientConnectionId = this.runtime.clientId;
+		assert(clientConnectionId !== undefined, 0xa59 /* Client connected without clientId */);
+		const currentClientToSessionValueState =
+			this.datastore["system:presence"].clientToSessionId[clientConnectionId];
+
+		const newMessage = {
 			sendTimestamp: Date.now(),
 			avgLatency: this.averageLatency,
 			// isComplete: false,
-			data,
+			data: {
+				// Always send current connection mapping for some resiliency against
+				// lost signals. This ensures that client session id found in `updates`
+				// (which is this client's client session id) is always represented in
+				// system workspace of recipient clients.
+				"system:presence": {
+					clientToSessionId: {
+						[clientConnectionId]: { ...currentClientToSessionValueState },
+					},
+				},
+				...this.queuedData,
+			},
 		} satisfies DatastoreUpdateMessage["content"];
-		this.runtime.submitSignal(datastoreUpdateMessageType, content);
+		this.queuedData = undefined;
+		this.timer.clearTimeout();
```

Comment:
Let's clear the timer first. It avoids having to think about od cases where there is an exception early that is caught and then timer fires later.

Reply:
"first" as in before the previous call? Or first thing in the sendQueuedData function?

Reply:
First thing in the function.